lib/Corres_Method.thy

```diff
@@ -0,0 +1,212 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+theory Corres_Method
+imports Corres_Cases ExtraCorres
+begin
+
+(* A proof method for automating simple steps in corres proofs.
+
+   While the method might solve some corres proofs completely, the purpose is to make simple
+   things more automatic, remove boilerplate, and to leave a proof state in which the user can make
+   more progress. The goal is not to provide full automation or deeper search.
+
+   The main idea is to repeatedly try to apply terminal [corres] rules after splitting off the head
+   of a bind/bindE statement on both sides of a corres goal. The method provides options for less
+   safe rules such as moving asserts to guards etc when the user knows that this is safe to do in
+   a particular instance.
+
+   See description at corres' method below for all parameters and options.
+*)
+
+section \<open>Goal predicates\<close>
+
+(* Succeed if the conclusion is a corres goal and also not purely schematic *)
+method is_corres = succeeds \<open>rule corres_inst\<close>, fails \<open>rule TrueI\<close>
+
+lemma no_fail_triv: "no_fail P f \<Longrightarrow> no_fail P f" .
+lemmas hoare_trivs = hoare_triv hoare_trivE hoare_trivE_R hoare_trivR_R no_fail_triv
+
+(* Succeed if the conclusion is a wp/no_fail goal and also not purely schematic*)
+method is_wp = succeeds \<open>rule hoare_trivs\<close>, fails \<open>rule TrueI\<close>
+
+lemmas hoare_post_False = hoare_pre_cont[where P="\<lambda>_. \<bottom>"]
+lemmas hoareE_post_False = hoare_FalseE[where Q="\<lambda>_. \<bottom>" and E="\<lambda>_. \<bottom>"]
```

Comment:
that naming consistency...

Reply:
yeah, we have this old old issue "wp lemmas could do with a mass rename"..